src/ng/parse.js

```diff
@@ -1963,7 +1966,14 @@ function $ParseProvider() {
         // If there is an interceptor, but no watchDelegate then treat the interceptor like
         // we treat filters - it is assumed to be a pure function unless flagged with $stateful
         fn.$$watchDelegate = inputsWatchDelegate;
-        fn.inputs = parsedExpression.inputs ? parsedExpression.inputs : [parsedExpression];
+        fn.inputs = (parsedExpression.inputs ? parsedExpression.inputs : [parsedExpression]).map(function(e) {
+              // Remove the isPure flag of inputs when it is not absolute because they are now wrapped in a
+              // potentially non-pure interceptor function.
```

Comment:
I think the comment a few lines up is using the wrong terminology. I think it should just say "it is assumed to be non-$stateful".  I'm not 100% sure if I'm using the "pure" terminology correct here, but I'll try to update the comments similar to #16015 which I think clears this up more. See: https://github.com/angular/angular.js/pull/16015/files#diff-780de070ede30180f6aedb6c5f7d57caR1964

To (try) summarizing:
* interceptors/filters can be `$stateful` in which case they must always be invoked to check for changes
* inputs can be `pure` in which case they only need to be shallow watched
* inputs passed into interceptors/filters are NOT pure because the interceptor/filter might inspect the content of them
* inputs that are pure in `$parse(exp)` might not be when doing `$parse(exp, interceptor)`, but they share the same `inputs` array which is what this ugly `.map` is for